source/opt/decoration_manager.cpp

```diff
@@ -231,31 +231,78 @@ std::vector<T> DecorationManager::InternalGetDecorationsFor(
-          break;
-        case SpvOpMemberDecorate:
-          if (inst->GetSingleWordInOperand(2) == decoration) {
-            f(*inst);
-          }
-          break;
-        case SpvOpDecorateId:
-          if (inst->GetSingleWordInOperand(1) == decoration) {
-            f(*inst);
+                                          std::function<void(const ir::Instruction&)> f) {
+  for (const ir::Instruction* inst : GetDecorationsFor(id, false)) {
+    switch (inst->opcode()) {
+      case SpvOpDecorate:
+        if (inst->GetSingleWordInOperand(1) == decoration) {
+          f(*inst);
+        }
+        break;
+      case SpvOpMemberDecorate:
+        if (inst->GetSingleWordInOperand(2) == decoration) {
+          f(*inst);
+        }
+        break;
+      case SpvOpDecorateId:
+        if (inst->GetSingleWordInOperand(1) == decoration) {
+          f(*inst);
+        }
+        break;
+      default:
+        assert(false && "Unexpected decoration instruction");
+    }
+  }
+}
+
+void DecorationManager::CloneDecorations(uint32_t from, uint32_t to) {
+  auto const decoration_list = id_to_decoration_insts_.find(from);
+  if (decoration_list == id_to_decoration_insts_.end()) return;
+  for (ir::Instruction* inst : decoration_list->second) {
+    switch (inst->opcode()) {
+      case SpvOpGroupDecorate:
+        // add |to| to list of decorated id's
+        inst->AddOperand(ir::Operand(spv_operand_type_t::SPV_OPERAND_TYPE_ID, {to}));
+        id_to_decoration_insts_[to].push_back(inst);
+        break;
+      case SpvOpGroupMemberDecorate: {
+        bool is_id = false; // we start with decoration group id, followed by target-id, literal, target-id...
+        bool to_add = false;
+        // for each (id == from), get literal and add (to, literal) as operands
+        inst->ForEachInOperand([&](uint32_t* operand) {
+          if (is_id) {
+            if (*operand == from) {
+              to_add = true;
+              inst->AddOperand(ir::Operand(spv_operand_type_t::SPV_OPERAND_TYPE_ID, {to}));
+            }
+            is_id = false;
+          } else { // is literal
+            if (to_add) {
+              to_add = false;
+              inst->AddOperand(ir::Operand(spv_operand_type_t::SPV_OPERAND_TYPE_LITERAL_INTEGER, {*operand}));
+            }
+            is_id = true;
           }
-          break;
-        default:
-          assert(false && "Unexpected decoration instruction");
+        });
+        assert(is_id && "mal-formed SpvOpGroupMemberDecorate encountered.");
+        id_to_decoration_insts_[to].push_back(inst);
+        break;
       }
+      case SpvOpDecorate:
+      case SpvOpMemberDecorate:
+      case SpvOpDecorateId: {
+        // simply clone decoration and change |target-id| to |to|
+        std::unique_ptr<ir::Instruction> newInst(inst->Clone());
+        newInst->SetInOperand(0, {to});
+        id_to_decoration_insts_[to].push_back(newInst.get());
+        module_->AddAnnotationInst(std::move(newInst));
```

Comment:
I prefer this idiom for adding instruction.  When we add an IRContext to the module, the module can take care of updating the analysis (including the decoration manager).

Reply:
So, in the other cases we better add a new instruction and tonop() the old one?

Reply:
Yes, I think that is a good way of doing it for now.